Oxidize Commutation Analysis

sbrandhsn opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary



### Details and comments
Part of #12208, oxidises `CommutationAnalysis`. This is based on PR #12959 (and thus #12550) and PR #12870 and shows promising runtime improvements in preliminary tests. When #12550 is merged, the interface to `CommutationChecker` needs a little thought, then this should be ready to go, in principle.

sbrandhsn committed:
init

sbrandhsn committed:
up

sbrandhsn committed:
lint

sbrandhsn committed:
.

sbrandhsn committed:
up

sbrandhsn committed:
before cache

sbrandhsn committed:
with cache

sbrandhsn committed:
correct

sbrandhsn committed:
cleaned up

sbrandhsn committed:
lint reno

sbrandhsn committed:
Update Cargo.lock

sbrandhsn committed:
Merge branch 'main' into oxidize-commutation-checker

sbrandhsn committed:
.

sbrandhsn committed:
Merge branch 'main' into oxidize-commutation-checker

sbrandhsn committed:
up

sbrandhsn committed:
.

sbrandhsn committed:
revert op

sbrandhsn committed:
.

sbrandhsn committed:
.

sbrandhsn committed:
.

sbrandhsn committed:
.

sbrandhsn committed:
Delete Cargo.lock

sbrandhsn committed:
.

sbrandhsn committed:
corrected string comparison

sbrandhsn committed:
removed Operator class from operation.rs

sbrandhsn committed:
.

sbrandhsn committed:
Apply suggestions from code review

Co-authored-by: Raynel Sanchez <87539502+raynelfss@users.noreply.github.com>

sbrandhsn committed:
comments from code review

kevinhartman committed:
Port DAGCircuit to Rust

This commit migrates the entirety of the `DAGCircuit` class to Rust. It
fully replaces the Python version of the class. The primary advantage
of this migration is moving from a Python space rustworkx directed graph
representation to a Rust space petgraph (the upstream library for
rustworkx) directed graph. Moving the graph data structure to rust
enables us to directly interact with the DAG directly from transpiler
passes in Rust in the future. This will enable a significant speed-up in
those transpiler passes. Additionally, this should also improve the
memory footprint as the DAGCircuit no longer stores `DAGNode`
instances, and instead stores a lighter enum NodeType, which simply
contains a `PackedInstruction` or the wire objects directly.

Internally, the new Rust-based `DAGCircuit` uses a `petgraph::StableGraph`
with node weights of type `NodeType` and edge weights of type `Wire`. The
NodeType enum contains variants for `QubitIn`, `QubitOut`, `ClbitIn`,
`ClbitOut`, and `Operation`, which should save us from all of the
`isinstance` checking previously needed when working with `DAGNode` Python
instances. The `Wire` enum contains variants `Qubit`, `Clbit`, and `Var`.

As the full Qiskit data model is not rust-native at this point while
all the class code in the `DAGCircuit` exists in Rust now, there are
still sections that rely on Python or actively run Python code via Rust
to function. These typically involve anything that uses `condition`,
control flow, classical vars, calibrations, bit/register manipulation,
etc. In the future as we either migrate this functionality to Rust or
deprecate and remove it this can be updated in place to avoid the use
of Python.

API access from Python-space remains in terms of `DAGNode` instances to
maintain API compatibility with the Python implementation. However,
internally, we convert to and deal in terms of NodeType. When the user
requests a particular node via lookup or iteration, we inflate an ephemeral
`DAGNode` based on the internal `NodeType` and give them that. This is very
similar to what was done in #10827 when porting CircuitData to Rust.

As part of this porting there are a few small differences to keep in
mind with the new Rust implementation of DAGCircuit. The first is that
the topological ordering is slightly different with the new DAGCircuit.
Previously, the Python version of `DAGCircuit` using a lexicographical
topological sort key which was basically `"0,1,0,2"` where the first
`0,1` are qargs on qubit indices `0,1` for nodes and `0,2` are cargs
on clbit indices `0,2`. However, the sort key has now changed to be
`(&[Qubit(0), Qubit(1)], &[Clbit(0), Clbit(2)])` in rust in this case
which for the most part should behave identically, but there are some
edge cases that will appear where the sort order is different. It will
always be a valid topological ordering as the lexicographical key is
used as a tie breaker when generating a topological sort. But if you're
relaying on the exact same sort order there will be differences after
this PR. The second is that a lot of undocumented functionality in the
DAGCircuit which previously worked because of Python's implicit support
for interacting with data structures is no longer functional. For
example, previously the `DAGCircuit.qubits` list could be set directly
(as the circuit visualizers previously did), but this was never
documented as supported (and would corrupt the DAGCircuit). Any
functionality like this we'd have to explicit include in the Rust
implementation and as they were not included in the documented public
API this PR opted to remove the vast majority of this type of
functionality.

The last related thing might require future work to mitigate is that
this PR breaks the linkage between `DAGNode` and the underlying
`DAGCirucit` object. In the Python implementation the `DAGNode` objects
were stored directly in the `DAGCircuit` and when an API method returned
a `DAGNode` from the DAG it was a shared reference to the underlying
object in the `DAGCircuit`. This meant if you mutated the `DAGNode` it
would be reflected in the `DAGCircuit`. This was not always a sound
usage of the API as the `DAGCircuit` was implicitly caching many
attributes of the DAG and you should always be using the `DAGCircuit`
API to mutate any nodes to prevent any corruption of the `DAGCircuit`.
However, now as the underlying data store for nodes in the DAG are
no longer the python space objects returned by `DAGCircuit` methods
mutating a `DAGNode` will not make any change in the underlying
`DAGCircuit`. This can come as quite the surprise at first, especially
if you were relying on this side effect, even if it was unsound.

It's also worth noting that 2 large pieces of functionality from
rustworkx are included in this PR. These are the new files
`rustworkx_core_vnext` and `dot_utils` which are rustworkx's VF2
implementation and its dot file generation. As there was not a rust
interface exposed for this functionality from rustworkx-core there was
no way to use these functions in rustworkx. Until these interfaces
added to rustworkx-core in future releases we'll have to keep these
local copies. The vf2 implementation is in progress in
Qiskit/rustworkx#1235, but `dot_utils` might make sense to keep around
longer term as it is slightly modified from the upstream rustworkx
implementation to directly interface with `DAGCircuit` instead of a
generic graph.

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>
Co-authored-by: Raynel Sanchez <87539502+raynelfss@users.noreply.github.com>
Co-authored-by: Elena Peña Tapia <57907331+ElePT@users.noreply.github.com>
Co-authored-by: Alexander Ivrii <alexi@il.ibm.com>
Co-authored-by: Eli Arbel <46826214+eliarbel@users.noreply.github.com>
Co-authored-by: John Lapeyre <jlapeyre@users.noreply.github.com>
Co-authored-by: Jake Lishman <jake.lishman@ibm.com>

mtreinish committed:
Update visual mpl circuit drawer references

Right now there is a bug in the matplotlib circuit visualizer likely
caused by the new `__eq__` implementation for `DAGOpNode` that didn't
exist before were some gates are missing from the visualization. In the
interest of unblocking this PR this commit updates the references for
these cases temporarily until this issue is fixed.

mtreinish committed:
Ensure DAGNode.sort_key is always a string

Previously the sort_key attribute of the Python space DAGCircuit was
incorrectly being set to `None` for rust generated node objects. This
was done as for the default path the sort key is determined from the
rust domain's representation of qubits and there is no analogous data in
the Python object. However, this was indavertandly a breaking API change
as sort_key is expected to always be a string. This commit adds a
default string to use for all node types so that we always have a
reasonable value that matches the typing of the class. A future step is
likely to add back the `dag` kwarg to the node types and generate the
string on the fly from the rust space data.

mtreinish committed:
Merge branch 'main' into oxidize-dag

mtreinish committed:
Make Python argument first in Param::eq and Param::is_close

The standard function signature convention for functions that take a
`py: Python` argument is to make the Python argument the first (or
second after `&self`). The `Param::eq` and `Param::is_close` methods
were not following this convention and had `py` as a later argument in
the signature. This commit corrects the oversight.

mtreinish committed:
Merge branch 'main' into oxidize-dag

mtreinish committed:
Fix merge conflict with #12943

With the recent merge with main we pulled in #12943 which conflicted
with the rust space API changes made in this PR branch. This commit
updates the usage to conform with the new interface introduced in this
PR.

mtreinish committed:
Add release notes and test for invalid args on apply methods

This commit adds several release notes to document this change. This
includes a feature note to describe the high level change and the user
facing benefit (mainly reduced memory consumption for DAGCircuits),
two upgrade notes to document the differences with shared references
caused by the new data structure, and a fix note documenting the fix
for how qargs and cargs are handled on `.apply_operation_back()` and
`.apply_operation_front()`. Along with the fix note a new unit test is
added to serve as a regression test so that we don't accidentally allow
adding cargs as qargs and vice versa in the future.

mtreinish committed:
Restore `inplace` argument functionality for substitute_node()

This commit restores the functionality of the `inplace` argument for
`substitute_node()` and restores the tests validating the object
identity when using the flag. This flag was originally excluded from
the implementation because the Rust representation of the dag is not
a shared reference with Python space and the flag doesn't really mean
the same thing as there is always a second copy of the data for Python
space now. The implementation here is cheating slighty as we're passed
in the DAG node by reference it relies on that reference to update the
input node at the same time we update the dag. Unlike the previous
Python implementation where we were updating the node in place and the
`inplace` argument was slightly faster because everything was done by
reference. The rust space data is still a compressed copy of the data
we return to Python so the `inplace` flag will be slightly more
inefficient as we need to copy to update the Python space representation
in addition to the rust version.

mtreinish committed:
Revert needless dict() cast on metadata in dag_to_circuit()

This commit removes an unecessary `dict()` cast on the `dag.metadata`
when setting it on `QuantumCircuit.metadata` in
`qiskit.converters.dag_to_circuit()`. This slipped in at some point
during the development of this PR and it's not clear why, but it isn't
needed so this removes it.

mtreinish committed:
Add code comment for DAGOpNode.__eq__ parameter checking

This commit adds a small inline code comment to make it clear why we
skip parameter comparisons in DAGOpNode.__eq__ for python ops. It might
not be clear why the value is hard coded to `true` in this case, as this
check is done via Python so we don't need to duplicate it in rust space.

mtreinish committed:
Raise a ValueError on DAGNode creation with invalid index

This commit adds error checking to the DAGNode constructor to raise a
PyValueError if the input index is not valid (any index < -1).
Previously this would have panicked instead of raising a user catchable
error.

mtreinish committed:
Use macro argument to set python getter/setter name

This commit updates the function names for `get__node_id` and
`set__node_id` method to use a name that clippy is happy with and
leverage the pyo3 macros to set the python space name correctly instead
of using the implicit naming rules.

mtreinish committed:
Remove Ord and PartialOrd derives from interner::Index

The Ord and PartialOrd traits were originally added to the Index struct
so they could be used for the sort key in lexicographical topological
sorting. However, that approach was abandonded during the development of
this PR and instead the expanded Qubit and Clbit indices were used
instead. This left the ordering traits as unnecessary on Index and
potentially misleading. This commit just opts to remove them as they're
not needed anymore.

kevinhartman committed:
Fix missing nodes in matplotlib drawer.

Previously, the change in equality for DAGNodes was causing nodes
to clobber eachother in the matplotlib drawer's tracking data
structures when used as keys to maps.

To fix this, we ensure that all nodes have a unique ID across
layers before constructing the matplotlib drawer. They actually
of course _do_ in the original DAG, but we don't really care
what the original IDs are, so we just make them up.

Writing to _node_id on a DAGNode may seem odd, but it exists
in the old Python API (prior to being ported to Rust) and
doesn't actually mutate the DAG at all since DAGNodes are
ephemeral.

mtreinish committed:
Revert "Update visual mpl circuit drawer references"

With the previous commit the bug in the matplotlib drawer causing the
images to diverge should be fixed. This commit reverts the change to the
reference images as there should be no difference now.

This reverts commit 1e4e6f386286b0b4e7f3ebd3f706f948dd707575.

mtreinish committed:
Update visual mpl circuit drawer references for control flow circuits

The earlier commit that "fixed" the drawers corrected the visualization
to match expectations in most cases. However after restoring the
references to what's on main several comparison tests with control flow
in the circuit were still failing. The failure mode looks similar to the
other cases, but across control flow blocks instead of at the circuit
level. This commit temporarily updates the references of these to the
state of what is generated currently to unblock CI. If/when we have a
fix this commit can be reverted.

sbrandhsn committed:
Apply suggestions from code review

Co-authored-by: Raynel Sanchez <87539502+raynelfss@users.noreply.github.com>

sbrandhsn committed:
code review

mtreinish committed:
Fully port Optimize1qGatesDecomposition to Rust

This commit builds off of #12550 and the other data model in Rust
infrastructure and migrates the Optimize1qGatesDecomposition pass to
operate fully in Rust. The full path of the transpiler pass now never
leaves rust until it has finished modifying the DAGCircuit. There is
still some python interaction necessary to handle parts of the data
model that are still in Python, mainly calibrations and parameter
expressions (for global phase). But otherwise the entirety of the pass
operates in rust now.

This is just a first pass at the migration here, it moves the pass to be
a single for loop in rust. The next steps here are to look at operating
the pass in parallel. There is no data dependency between the
optimizations being done by the pass so we should be able to the
throughput of the pass by leveraging multithreading to handle each run
in parallel. This commit does not attempt this though, because of the
Python dependency and also the data structures around gates and the
dag aren't really setup for multithreading yet and there likely will
need to be some work to support that (this pass is a good candidate to
work through the bugs on that).

Part of #12208

mtreinish committed:
Fix edge cases in DAGOpNode.__eq__

This commit fixes a couple of edge cases in DAGOpNode.__eq__ method
around the python interaction for the method. The first is that in
the case where we had python object parameter types for the gates we
weren't comparing them at all. This is fixed so we use python object
equality for the params in this case. Then we were dropping the error
handling in the case of using python for equality, this fixes it to
return the error to users if the equality check fails. Finally a comment
is added to explain the expected use case for `DAGOpNode.__eq__` and why
parameter checking is more strict than elsewhere.

mtreinish committed:
Remove Param::add() for global phase addition

This commit removes the Param::add() method and instead adds a local
private function to the `dag_circuit` module for doing global phase
addition. Previously the `Param::add()` method was used solely for
adding global phase in `DAGCircuit` and it took some shortcuts knowing
that context. This made the method implementation ill suited as a
general implementation.

kevinhartman committed:
More complete fix for matplotlib drawer.

kevinhartman committed:
Revert "Update visual mpl circuit drawer references for control flow circuits"

This reverts commit 9a6f9536a3a7412d19a9fd9bbd761825c9a53d0f.

mtreinish committed:
Unify rayon versions in workspace

kevinhartman committed:
Remove unused _GLOBAL_NID.

mtreinish committed:
Merge branch 'main' into oxidize-dag

mtreinish committed:
Use global monotonic ID counter for ids in drawer

The fundamental issue with matplotlib visualizations of control flow is
that locally in the control flow block the nodes look the same but are
stored in an outer circuit dictionary. If the gates are the same and on
the same qubits and happen to have the same node id inside the different
control flow blocks the drawer would think it's already drawn the node
and skip it incorrectly. The previous fix for this didn't go far enough
because it wasn't accounting for the recursive execution of the drawer
for inner blocks (it also didn't account for LayerSpoolers of the same
length).

mtreinish committed:
Merge remote-tracking branch 'origin/main' into oxidize-dag

mtreinish committed:
Merge branch 'main' into oxidize-dag

kevinhartman committed:
Remove unused BitData iterator stuff.

sbrandhsn committed:
remove with_gil in favor of passing python tokens as params

sbrandhsn committed:
Merge branch 'pr/12959' into oxidize-commutative-analysis

sbrandhsn committed:
Apply suggestions from code review

Co-authored-by: Raynel Sanchez <87539502+raynelfss@users.noreply.github.com>

sbrandhsn committed:
fmt

sbrandhsn committed:
python serialization

sbrandhsn committed:
deprecation

sbrandhsn committed:
Update commutation_checker.py

sbrandhsn committed:
heh

sbrandhsn committed:
init

sbrandhsn committed:
let Pytuple collect

sbrandhsn committed:
lint

sbrandhsn committed:
Merge branch 'main' into oxidize-commutation-checker

mtreinish committed:
Merge remote-tracking branch 'origin/main' into oxidize-commutation-checker

Cryoris committed:
First set of comments

- use Qubit/Clbit
- more info on unsafe
- update reno
- use LazySet less
- use OperationRef, avoid CircuitInstruction creation

Cryoris committed:
Second part

- clippy
- no BigInt
- more comments

Cryoris committed:
Matrix speed & fix string sort

-- could not use op.name() directly since sorted differently than Python, hence it's back to BigInt

Cryoris committed:
have the Python implementation use Rust

Cryoris committed:
lint & tools

Cryoris committed:
remove unsafe blocks

Cryoris committed:
One more try to avoid segfaulty windows

-- if that doesn't work maybe revert the change the the Py CommChecker uses Rust

Cryoris committed:
Merge branch 'main' into oxidize-commutation-checker

Cryoris committed:
Original version

Co-authored-by: Sebastian Brandhofer <148463728+sbrandhsn@users.noreply.github.com>

Cryoris committed:
Sync with updated CommutationChecker

todo: shouldn't make the qubits interner public

Cryoris committed:
Debug: disable cache

trying to figure out why the windows CI fails (after being unable to locally reproduce we're using CI with a reduced set of tests)

Cryoris committed:
... second try

mtreinish committed:
Update crates/accelerate/src/commutation_checker.rs

Co-authored-by: Raynel Sanchez <87539502+raynelfss@users.noreply.github.com>

mtreinish committed:
Restore azure config

mtreinish committed:
Merge remote-tracking branch 'origin/main' into oxidize-commutation-checker

mtreinish committed:
Remove unused import

mtreinish committed:
Revert "Debug: disable cache"

This reverts commit c564b806c9ff31d5faa6aa3f02c18cfe6a5ff2c5.

mtreinish committed:
Don't overallocate cache

We were allocating a the cache hashmap with a capacity for max cache
size entries every time we instantiated a new CommutationChecker. The
max cache size is 1 million. This meant we were allocating 162MB
everytime CommutationChecker.__new__ was called, which includes each
time we instantiate it manually (which happens once on import), the
CommutationAnalysis pass gets instantiated (twice per preset pass
manager created with level 2 or 3), or a commutation checker instance is
pickle deserialized. This ends up causing a fairly large memory
regression and is the source of the CI failures on windows.

Co-authored-by: Jake Lishman <jake.lishman@ibm.com>

mtreinish committed:
Cleanup parameter key type to handle edge conditions better

This commit cleans up the ParameterKey type and usage to make it handle
edge conditions better. The first is that the type just doesn't do the
right thing for NaN, -0, or the infinities. Canonicalization is added
for hash on -0 and the only constructor of the newtype adds a runtime
guard against NaN and inifinity (positive or negative) to avoid that
issue. The approach only makes sense as the cache is really there to
guard us against unnecessary re-computing when we reiterate over the
circuit > 1 time and nothing has changed for gates. Otherwise comparing
floats like done in this PR does would not be a sound or an effective
approach.

mtreinish committed:
Remove unnecessary cache hit rate tracking

mtreinish committed:
Undo test assertion changes

mtreinish committed:
Undo unrelated test changes

mtreinish committed:
Merge remote-tracking branch 'origin/main' into oxidize-commutation-checker

mtreinish committed:
Undo pending deprecation and unify commutation classes

This commit removes the pending deprecation decorator from the python
class definition as the Python class just internally is using the rust
implementation now. This also removes directly using the rust
implementation for the standard commutation library global as using the
python class is exactly the same now.

We can revisit if there is anything we want to deprecate and remove in
2.0 in a follow up PR. Personally, I think the cache management methods
are all we really want to remove as the cache should be an internal
implementation detail and not part of the public interface.

mtreinish committed:
Undo gha config changes

mtreinish committed:
Make serialization explicit

This commit makes the pickling of cache entries explicit. Previously it
was relying on conversion traits which hid some of the complexity but
this uses a pair of conversion functions instead.

mtreinish committed:
Remove stray SAFETY comment

mtreinish committed:
Remove ddt usage from the tests

Now that the python commutation checker and the rust commutation checker
are the same thing the ddt parameterization of the commutation checker
tests was unecessary duplication. This commit removes the ddt usage to
restore having a single run of all the tests.

mtreinish committed:
Update release note

mtreinish committed:
Fix CommutationChecker class import

mtreinish committed:
Remove invalid test assertion for no longer public attribute

Cryoris committed:
Ray's review comments

Co-authored-by: Raynel Sanchez <raynelfss@hotmail.com>

Cryoris committed:
Handle ``atol/rtol``, more error propagation

Cryoris committed:
Merge branch 'oxidize-commutation-checker' into rust/commutation-analysis

Cryoris committed:
update to latest changes in commchecker

Cryoris committed:
Merge branch 'rust/commutation-analysis' into oxidize-commutative-analysis

Cryoris committed:
Merge branch 'main' into oxidize-commutative-analysis

Cryoris committed:
fix merge conflict remnants

Cryoris committed:
re-use expensive quantities

such as the relative placement and the parameter hash

Cryoris committed:
Merge branch 'oxidize-commutation-checker' into oxidize-commutative-analysis

Cryoris committed:
Merge branch 'main' into oxidize-commutative-analysis

Cryoris committed:
add missing header

mtreinish commented on crates/accelerate/src/commutation_analysis.rs:
```diff
@@ -0,0 +1,179 @@
+use qiskit_circuit::dag_circuit::{DAGCircuit, NodeType, Wire};
+use rustworkx_core::petgraph::stable_graph::NodeIndex;
+
+type AIndexSet<T> = IndexSet<T, BuildHasherDefault<AHasher>>;
```
This can be expressed a bit more succinctly with:

```suggestion
type AIndexSet<T> = IndexSet<T, ::ahash::RandomState>;
```

That being said I'm not sure I'm a fan of using a type alias for this instead of just using `IndexSet<NodeIndex, RandomState>` (assuming a use `ahash::RandomState` is added) where we need to define a new indexset.

mtreinish commented on crates/accelerate/src/commutation_analysis.rs:
```diff
@@ -0,0 +1,179 @@
+                                let qargs1 = dag.qargs_interner.get(packed_inst0.qubits);
+                                let qargs2 = dag.qargs_interner.get(packed_inst1.qubits);
+                                let cargs1 = dag.cargs_interner.get(packed_inst0.clbits);
+                                let cargs2 = dag.cargs_interner.get(packed_inst1.clbits);
```
```suggestion
                                let qargs1 = dag.get_qargs(packed_inst0.qubits);
                                let qargs2 = dag.get_qargs((packed_inst1.qubits);
                                let cargs1 = dag.get_cargs(packed_inst0.clbits);
                                let cargs2 = dag.get_cargs(packed_inst1.clbits);
```

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -231,9 +231,9 @@ pub struct DAGCircuit {
 
     /// The cache used to intern instruction qargs.
-    qargs_interner: Interner<[Qubit]>,
+    pub qargs_interner: Interner<[Qubit]>,
```
There are `get_qargs` and `get_cargs` methods on `main` now so you don't need to make this public:

```suggestion
    qargs_interner: Interner<[Qubit]>,
```

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -231,9 +231,9 @@ pub struct DAGCircuit {
+    pub qargs_interner: Interner<[Qubit]>,
     /// The cache used to intern instruction cargs.
-    cargs_interner: Interner<[Clbit]>,
+    pub cargs_interner: Interner<[Clbit]>,
```
```suggestion
    cargs_interner: Interner<[Clbit]>,
```

mtreinish commented on crates/accelerate/src/commutation_analysis.rs:
```diff
@@ -0,0 +1,179 @@
+                                let params2 = match packed_inst1.params.as_ref() {
+                                    Some(params) => params,
+                                    None => &empty_params,
+                                };
```
You can just use `params_view()` for this:

```suggestion
                                let op1 = packed_inst0.op.view();
                                let op2 = packed_inst1.op.view();
                                let params1 = packed_inst0.params_view();
                                let params2 = packed_inst1.params_view();
```

Cryoris replied:
ahh perfect -- thanks!

Cryoris committed:
gentler error handling

mtreinish reviewed (commented):
This is looking good thanks for doing this. I left some small mostly mechanical comments inline. I still want to review the pass logic to double check it all matches up exactly. Especially as there seem to be a fair amount of `panic!()` calls which I want to double check and make sure we're not masking potential user actionable error conditions (or indicate a logic issue).

Cryoris committed:
review comments & more docs

mtreinish commented on crates/accelerate/src/commutation_analysis.rs:
```diff
@@ -0,0 +1,197 @@
+    let mut commutation_set: HashMap<Wire, CommutingNodes> = HashMap::new();
+    let mut node_indices: HashMap<(NodeIndex, Wire), usize> = HashMap::new();
+
+    let max_num_qubits = 3;
```
Not that it matters, but I'd have done this with:

```rust
const MAX_NUM_QUBITS: u32 = 3;
```

outside the function definition because it is never changed. But realistically the compiler will likely end up with the same generated code in both cases so it doesn't matter.

Cryoris replied:
Still consistency is nice -- I'll move it outside 👍🏻

mtreinish commented on crates/accelerate/src/commutation_analysis.rs:
```diff
@@ -0,0 +1,197 @@
+                commutation_entry.len() - 1,
+            );
+        }
+    }
```
This is a probably a good candidate for parallelization. We can naively make this a parallel iterator for the outer loop like:

```rust
0..dag.num_qubits().into_par_iter().map(|qubit| Wire::Qubit(Qubit(qubit as u32))).for_each(|wire| {
```

We'll probably have to play some games around synchronization, although we can probably change it into an iterator collector. But we should do this in a follow up because it might require some deeper changes (as the compiler points them out).

Cryoris replied:
Yeah good point -- I'll have a look in parallel 🙂 I was trying to propagate the Result coming from `commute_inner` but if we collect then we can still do that 👍🏻

mtreinish commented on crates/accelerate/src/commutation_analysis.rs:
```diff
@@ -0,0 +1,197 @@
+                for prev_gate_idx in last.iter() {
+                    // if the node is an input/output node, they do not commute, so we only
+                    // continue if the nodes are operation nodes
+                    if let (NodeType::Operation(packed_inst0), NodeType::Operation(packed_inst1)) =
+                        (&dag.dag[current_gate_idx], &dag.dag[*prev_gate_idx])
```
Can `current_gate_idx` ever point to a non-op node? Not that it really matters because we have need to get the inner packed instruction, I was just wondering from a logic PoV.

Cryoris replied:
yep it even has to, since the commutation sets also contain indices for the input and output nodes (we iterate over `dag.nodes_on_wire(..., only_ops: false)`)

mtreinish commented on crates/accelerate/src/commutation_analysis.rs:
```diff
@@ -0,0 +1,197 @@
+// custom types: IndexSet allows to iterate over the elements in insertion
+// order, but uses the std hasher, which is slower than ahash
+type AIndexSet<T> = IndexSet<T, ::ahash::RandomState>;
+type CommutingNodes = Vec<AIndexSet<NodeIndex>>;
```
I'm wondering about the choice of `IndexSet` here is this just for deduplication? The original python code doesn't seem to use a set for anything.

Cryoris replied:
@sbrandhsn motivation was to have a fast lookup (we check if the current gate index is already in the commutation set). I can check what the performance is with a normal vector 🙂

sbrandhsn replied:
I was hoping for a runtime speedup coming from a faster execution of line 79 with a set instead of a list :-) I have not benchmarked this though. It might not be a sane optimization as we probably would only get a runtime speedup when we have large sets of commuting gates. It depends on the input circuit but we might assume that large commuting set do not occur in practice. Besides, I thought if the variable name says that the variable is a set, it should be a set instead of a list. :-D I did not consider deduplication.

mtreinish replied:
Switched to a `Vec` in: https://github.com/Qiskit/qiskit/pull/12995/commits/70735707d1a1490502d757528399f6b461fbdbda

mtreinish commented on crates/accelerate/src/commutation_analysis.rs:
```diff
@@ -0,0 +1,197 @@
+    // Then we add the {(node, wire): index} dictionary
+    for ((node_index, wire), index) in node_indices {
+        // we could cache the py_wires to avoid this match and the python object creation,
+        // but this didn't make a noticable difference in runtime
```
Realistically the `BitData` you're accessing here is the cache. It's basically just doing a vec lookup and returning the python object.

That being said I'm wondering can wire here be anything besides a `Qubit`? Not that it really matters because this code is correct and will handle the other cases as expected.

Cryoris replied:
No you're right, it cannot, since we explicitly only iterate over `0..dag.num_qubits().map(|i| Wire::Qubit(i))`. We can change the code to expect a Qubit, that might be a bit cleaner to the readers

sbrandhsn replied:
I was wondering what happens if a dag has non-continuous qubits, i.e. gaps in the set of qubits such as qubits 0...3 and 6..10 with no qubits at index 4, 5 - can this happen and was this supported in the Python implementation?

mtreinish replied:
The indices can't ever be non-continuous in Rust or Python. They're literally the index in the `list`/`Vec` for `dag.qubits`. So if you remove a qubit everything has a shift penalty.  The rust code just makes this a bit more explicit than it ever was in Python because there was a layer of indirection around it in Python with the bit objects.

mtreinish reviewed (commented):
Overall this LGTM, after the most recent update the code looks good to me. I had a few questions inline nothing really blocking. But at least for the data structure I'd like to know the rationale behind using a `HashMap<Wire, Vec<IndexSet<NodeIndex>>` as the type when in python it was all lists. I can believe using a `Vec<IndexSet>` as the weight here is the better choice but I'd like to know the rationale before approving.

Cryoris committed:
Use vec over IndexSet + clippy

- vec<vec> is slightly faster than vec<indexset>
- add custom types to satisfies clippy's complex type complaint
- don't handle Clbit/Var

mtreinish commented:
I ran ASV comparing this against main to see what the performance looks like and it's really good:
```
Benchmarks that have improved:

| Change   | Before [dff9e812]    | After [70735707] <oxidize-commutative-analysis>   |   Ratio | Benchmark (Parameter)                                             |
|----------|----------------------|---------------------------------------------------|---------|-------------------------------------------------------------------|
| -        | 241±2ms              | 217±0.6ms                                         |    0.9  | utility_scale.UtilityScaleBenchmarks.time_bv_100('cx')            |
| -        | 250±0.8ms            | 224±2ms                                           |    0.9  | utility_scale.UtilityScaleBenchmarks.time_bv_100('cz')            |
| -        | 4.08±0.02s           | 3.64±0.01s                                        |    0.89 | utility_scale.UtilityScaleBenchmarks.time_qv('ecr')               |
| -        | 722±6ms              | 646±3ms                                           |    0.89 | utility_scale.UtilityScaleBenchmarks.time_square_heisenberg('cz') |
| -        | 1.38±0.01s           | 1.20±0.01s                                        |    0.87 | utility_scale.UtilityScaleBenchmarks.time_qaoa('ecr')             |
| -        | 1.75±0.01s           | 1.46±0.01s                                        |    0.83 | utility_scale.UtilityScaleBenchmarks.time_qaoa('cz')              |
| -        | 6.89±0.02s           | 3.61±0.03s                                        |    0.52 | utility_scale.UtilityScaleBenchmarks.time_qft('cz')               |
| -        | 6.42±0.02s           | 3.18±0.02s                                        |    0.5  | utility_scale.UtilityScaleBenchmarks.time_qft('ecr')              |
| -        | 5.53±0.01s           | 2.46±0.02s                                        |    0.44 | utility_scale.UtilityScaleBenchmarks.time_qft('cx')               |
| -        | 84.8±1ms             | 33.1±0.3ms                                        |    0.39 | passes.PassBenchmarks.time_commutation_analysis(14, 1024)         |
| -        | 133±0.8ms            | 48.7±0.2ms                                        |    0.37 | passes.PassBenchmarks.time_commutation_analysis(20, 1024)         |
| -        | 25.9±0.8ms           | 9.55±0.1ms                                        |    0.37 | passes.PassBenchmarks.time_commutation_analysis(5, 1024)          |
| -        | 36.6±0.2ms           | 13.3±0.2ms                                        |    0.36 | utility_scale.UtilityScaleBenchmarks.time_bvlike('cz')            |
| -        | 36.6±0.2ms           | 13.0±0.07ms                                       |    0.35 | utility_scale.UtilityScaleBenchmarks.time_bvlike('cx')            |
| -        | 37.1±0.1ms           | 13.1±0.07ms                                       |    0.35 | utility_scale.UtilityScaleBenchmarks.time_bvlike('ecr')           |

Benchmarks that have stayed the same:

| Change   | Before [dff9e812]    | After [70735707] <oxidize-commutative-analysis>   | Ratio   | Benchmark (Parameter)                                                                                           |
|----------|----------------------|---------------------------------------------------|---------|-----------------------------------------------------------------------------------------------------------------|
|          | 5.40±0.03s           | 4.81±0.03s                                        | ~0.89   | utility_scale.UtilityScaleBenchmarks.time_qv('cz')                                                              |
|          | 0                    | 0                                                 | n/a     | utility_scale.UtilityScaleBenchmarks.track_bvlike_depth('cx')                                                   |
|          | 0                    | 0                                                 | n/a     | utility_scale.UtilityScaleBenchmarks.track_bvlike_depth('cz')                                                   |
|          | 0                    | 0                                                 | n/a     | utility_scale.UtilityScaleBenchmarks.track_bvlike_depth('ecr')                                                  |
|          | 9.83±0.07ms          | 10.3±0.08ms                                       | 1.05    | utility_scale.UtilityScaleBenchmarks.time_parse_qaoa_n100('ecr')                                                |
|          | 34.4±0.1ms           | 36.0±0.1ms                                        | 1.05    | utility_scale.UtilityScaleBenchmarks.time_parse_square_heisenberg_n100('cx')                                    |
|          | 34.3±0.2ms           | 36.0±0.1ms                                        | 1.05    | utility_scale.UtilityScaleBenchmarks.time_parse_square_heisenberg_n100('cz')                                    |
|          | 171±0.8ms            | 177±0.7ms                                         | 1.04    | passes.MultipleBasisPassBenchmarks.time_optimize_1q_commutation(5, 1024, ['rx', 'ry', 'rz', 'r', 'rxx', 'id'])  |
|          | 184±1ms              | 191±0.3ms                                         | 1.04    | passes.MultipleBasisPassBenchmarks.time_optimize_1q_commutation(5, 1024, ['rz', 'x', 'sx', 'cx', 'id'])         |
|          | 166±0.4ms            | 173±0.7ms                                         | 1.04    | passes.MultipleBasisPassBenchmarks.time_optimize_1q_commutation(5, 1024, ['u', 'cx', 'id'])                     |
|          | 9.87±0.05ms          | 10.3±0.03ms                                       | 1.04    | utility_scale.UtilityScaleBenchmarks.time_parse_qaoa_n100('cx')                                                 |
|          | 9.92±0.05ms          | 10.3±0.09ms                                       | 1.04    | utility_scale.UtilityScaleBenchmarks.time_parse_qaoa_n100('cz')                                                 |
|          | 106±0.3ms            | 110±0.5ms                                         | 1.04    | utility_scale.UtilityScaleBenchmarks.time_parse_qft_n100('cx')                                                  |
|          | 107±0.6ms            | 111±0.5ms                                         | 1.04    | utility_scale.UtilityScaleBenchmarks.time_parse_qft_n100('cz')                                                  |
|          | 106±0.8ms            | 110±0.6ms                                         | 1.04    | utility_scale.UtilityScaleBenchmarks.time_parse_qft_n100('ecr')                                                 |
|          | 34.5±0.5ms           | 35.8±0.2ms                                        | 1.04    | utility_scale.UtilityScaleBenchmarks.time_parse_square_heisenberg_n100('ecr')                                   |
|          | 554±2ms              | 570±2ms                                           | 1.03    | passes.MultipleBasisPassBenchmarks.time_optimize_1q_commutation(14, 1024, ['rx', 'ry', 'rz', 'r', 'rxx', 'id']) |
|          | 582±2ms              | 599±2ms                                           | 1.03    | passes.MultipleBasisPassBenchmarks.time_optimize_1q_commutation(14, 1024, ['rz', 'x', 'sx', 'cx', 'id'])        |
|          | 549±2ms              | 563±4ms                                           | 1.02    | passes.MultipleBasisPassBenchmarks.time_optimize_1q_commutation(14, 1024, ['u', 'cx', 'id'])                    |
|          | 908±4ms              | 925±2ms                                           | 1.02    | passes.MultipleBasisPassBenchmarks.time_optimize_1q_commutation(20, 1024, ['rx', 'ry', 'rz', 'r', 'rxx', 'id']) |
|          | 947±2ms              | 966±3ms                                           | 1.02    | passes.MultipleBasisPassBenchmarks.time_optimize_1q_commutation(20, 1024, ['rz', 'x', 'sx', 'cx', 'id'])        |
|          | 896±6ms              | 914±3ms                                           | 1.02    | passes.MultipleBasisPassBenchmarks.time_optimize_1q_commutation(20, 1024, ['u', 'cx', 'id'])                    |
|          | 3.70±0.1s            | 3.73±0.01s                                        | 1.01    | utility_scale.UtilityScaleBenchmarks.time_circSU2('cx')                                                         |
|          | 395                  | 395                                               | 1.00    | utility_scale.UtilityScaleBenchmarks.track_bv_100_depth('cx')                                                   |
|          | 397                  | 397                                               | 1.00    | utility_scale.UtilityScaleBenchmarks.track_bv_100_depth('cz')                                                   |
|          | 397                  | 397                                               | 1.00    | utility_scale.UtilityScaleBenchmarks.track_bv_100_depth('ecr')                                                  |
|          | 300                  | 300                                               | 1.00    | utility_scale.UtilityScaleBenchmarks.track_circSU2_depth('cx')                                                  |
|          | 300                  | 300                                               | 1.00    | utility_scale.UtilityScaleBenchmarks.track_circSU2_depth('cz')                                                  |
|          | 300                  | 300                                               | 1.00    | utility_scale.UtilityScaleBenchmarks.track_circSU2_depth('ecr')                                                 |
|          | 1607                 | 1607                                              | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qaoa_depth('cx')                                                     |
|          | 1622                 | 1622                                              | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qaoa_depth('cz')                                                     |
|          | 1622                 | 1622                                              | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qaoa_depth('ecr')                                                    |
|          | 1954                 | 1954                                              | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qft_depth('cx')                                                      |
|          | 1954                 | 1954                                              | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qft_depth('cz')                                                      |
|          | 1954                 | 1954                                              | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qft_depth('ecr')                                                     |
|          | 2709                 | 2709                                              | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qv_depth('cx')                                                       |
|          | 2709                 | 2709                                              | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qv_depth('cz')                                                       |
|          | 2709                 | 2709                                              | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qv_depth('ecr')                                                      |
|          | 462                  | 462                                               | 1.00    | utility_scale.UtilityScaleBenchmarks.track_square_heisenberg_depth('cx')                                        |
|          | 462                  | 462                                               | 1.00    | utility_scale.UtilityScaleBenchmarks.track_square_heisenberg_depth('cz')                                        |
|          | 462                  | 462                                               | 1.00    | utility_scale.UtilityScaleBenchmarks.track_square_heisenberg_depth('ecr')                                       |
|          | 3.80±0.06s           | 3.73±0.1s                                         | 0.98    | utility_scale.UtilityScaleBenchmarks.time_circSU2('cz')                                                         |
|          | 3.91±0.02s           | 3.85±0.2s                                         | 0.98    | utility_scale.UtilityScaleBenchmarks.time_circSU2('ecr')                                                        |
|          | 2.87±0.02s           | 2.74±0.02s                                        | 0.95    | utility_scale.UtilityScaleBenchmarks.time_qv('cx')                                                              |
|          | 450±6ms              | 423±5ms                                           | 0.94    | utility_scale.UtilityScaleBenchmarks.time_square_heisenberg('cx')                                               |
|          | 613±3ms              | 572±2ms                                           | 0.93    | utility_scale.UtilityScaleBenchmarks.time_square_heisenberg('ecr')                                              |
|          | 241±0.5ms            | 222±1ms                                           | 0.92    | utility_scale.UtilityScaleBenchmarks.time_bv_100('ecr')                                                         |
|          | 810±6ms              | 743±5ms                                           | 0.92    | utility_scale.UtilityScaleBenchmarks.time_qaoa('cx')                                                            |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

mtreinish commented on qiskit/transpiler/passes/optimization/commutation_analysis.py:
```diff
@@ -33,6 +32,13 @@ def __init__(self, *, _commutation_checker=None):
+            # is passed, defer to the internally held Rust implementation instead. This is needed
+            # as the ``analyze_commutations`` function expects the Rust implementation.
+            _commutation_checker = _commutation_checker.cc
+
```
We can remove this condition, we didn't deprecate the Python side and just unified the implementations. I don't think we're ever be getting the inner rust object directly anymore.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/12995/commits/68f4f4b51748e1e581e11d8e636aceb8df59fe2b

mtreinish reviewed (commented):
This LGTM, one super small thing in the Python code that is out of date with what got merged in for the commutation checker. It's not a bug but we just have an extra condition and comment that aren't needed anymore.

mtreinish committed:
Simplify python class construction

Since this PR was first written the split between the python side and
rust side of the CommutationChecker class has changed so that there are
no longer separate classes anymore. The implementations are unified and
the python space class just wraps an inner rust object. However, the
construction of the CommutationAnalysis pass was still written assuming
there was the possibility to get either a rust or Python object. This
commit fixes this and the type change on the `comm_checker` attribute by
removing the unnecessary logic.

mtreinish reviewed (approved):
I fixed that last little nit and now this should be ready to merge. Thanks @sbrandhsn and @Cryoris for doing this!
